crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -191,46 +228,507 @@ pub fn run_enter_schedule<S: States>(world: &mut World) {
+                        world.send_event(StateTransitionEvent {
+                            before: exited.clone(),
+                            after: entered.clone(),
+                        });
+                        world
+                            .try_run_schedule(ComputeDependantStates::<S>::default())
+                            .ok();
+
+                        let mut state_transition_schedules =
+                            world.get_resource_or_insert_with(StateTransitionSchedules::default);
+                        state_transition_schedules
+                            .exit_schedules
+                            .insert(OnExit(exited.clone()).intern());
+                        state_transition_schedules.transition_schedules.insert(
+                            OnTransition {
+                                from: exited,
+                                to: entered.clone(),
+                            }
+                            .intern(),
+                        );
+                        state_transition_schedules
+                            .enter_schedules
+                            .insert(OnEnter(entered.clone()).intern());
+                    }
+                }
+                None => {
+                    world.insert_resource(State(entered.clone()));
+                    world
+                        .try_run_schedule(ComputeDependantStates::<S>::default())
+                        .ok();
+
+                    let mut state_transition_schedules =
+                        world.get_resource_or_insert_with(StateTransitionSchedules::default);
+                    state_transition_schedules
+                        .enter_schedules
+                        .insert(OnEnter(entered.clone()).intern());
+                }
+            };
+        }
+        None => {
+            if let Some(resource) = world.remove_resource::<State<S>>() {
+                world
+                    .try_run_schedule(ComputeDependantStates::<S>::default())
+                    .ok();
+
+                let mut state_transition_schedules =
+                    world.get_resource_or_insert_with(StateTransitionSchedules::default);
+                state_transition_schedules
+                    .exit_schedules
+                    .insert(OnExit(resource.0).intern());
+            }
+        }
+    }
+}
+
+fn prepare_state_transitions(world: &mut World) {
+    world.insert_resource(StateTransitionSchedules::default());
+}
+
+fn execute_state_transition_schedules(world: &mut World) {
+    if let Some(schedules) = world.remove_resource::<StateTransitionSchedules>() {
+        for schedule in schedules.exit_schedules {
+            let _ = world.try_run_schedule(schedule);
+        }
+
+        for schedule in schedules.transition_schedules {
+            let _ = world.try_run_schedule(schedule);
+        }
+
+        for schedule in schedules.enter_schedules {
+            let _ = world.try_run_schedule(schedule);
+        }
```

Comment:
I think I preferred the old convention, because it ordered computed states correctly in reference to states they were derived from.
Here the execution is ambiguous between e.g. `OnExit` of various states.
But I think it's fine for now, the internals can be changed later.

Reply:
Discussed in discord, but for posterity I wanted to provide the record here: due to the way you can have multiple sources causing changes to a computed state, we would risk having ambiguous ordering between those. This sets up a consistent expectation - by the time the transition schedules happen, things have been processed and stored.